storage/scylla_api.c

```diff
@@ -0,0 +1,635 @@
+                                              value bigint, \
+                                              timestamp bigint, \
+                                              trunk blob, \
+                                              branch blob, \
+                                              PRIMARY KEY (bundle, address, transaction));") != CASS_OK) {
+    ret = SC_STORAGE_CASS_ERROR;
+    goto exit;
+  }
+exit:
+  return ret;
+}
+
+static status_t create_edge_table(CassSession* session) {
+  status_t ret = SC_OK;
+  if (execute_query(session, "DROP TABLE IF EXISTS permanent.edgeTable;") != CASS_OK) {
+    ret = SC_STORAGE_CASS_ERROR;
+    goto exit;
+  }
+  if (execute_query(session,
+                    "CREATE TABLE IF NOT EXISTS permanent.edgeTable (edge blob, \
+                                              bundle blob, \
+                                              address blob, \
+                                              transaction blob, \
+                                              PRIMARY KEY (edge, bundle, address, transaction));") != CASS_OK) {
+    ret = SC_STORAGE_CASS_ERROR;
+    goto exit;
+  }
+exit:
+  return ret;
+}
+status_t init_scylla(CassCluster* cluster, CassSession* session, char* hosts, bool is_need_create_table) {
+  status_t ret = SC_OK;
+  /* Add contact points */
+  cluster = create_cluster(hosts);
+
+  if (connect_session(session, cluster) != CASS_OK) {
+    SCYLLA_LOG_ERROR("%s\n", "connect Scylla cluster fail");
+    ret = SC_STORAGE_CONNECT_FAIL;
+    cass_cluster_free(cluster);
+    cass_session_free(session);
+    goto exit;
+  }
+  if (is_need_create_table == true) {
+    if ((ret = create_bundle_table(session)) != SC_OK) {
+      goto exit;
+    }
+    if ((ret = create_edge_table(session)) != SC_OK) {
+      goto exit;
+    }
+  }
+
+exit:
+  return ret;
+}
+static CassStatement* ret_insert_bundle_statement(const CassPrepared* prepared, const Transaction* transaction) {
+  CassStatement* statement = NULL;
+  statement = cass_prepared_bind(prepared);
+  cass_statement_bind_bytes_by_name(statement, "bundle", transaction->bundle, NUM_FLEX_TRITS_BUNDLE);
+  cass_statement_bind_bytes_by_name(statement, "address", transaction->address, NUM_FLEX_TRITS_ADDRESS);
+  cass_statement_bind_bytes_by_name(statement, "transaction", transaction->transaction, NUM_FLEX_TRITS_HASH);
+  cass_statement_bind_string_by_name(statement, "message", transaction->message);
+  cass_statement_bind_int64_by_name(statement, "value", transaction->value);
+  cass_statement_bind_int64_by_name(statement, "timestamp", transaction->timestamp);
+  cass_statement_bind_bytes_by_name(statement, "trunk", transaction->trunk, NUM_FLEX_TRITS_TRUNK);
+  cass_statement_bind_bytes_by_name(statement, "branch", transaction->branch, NUM_FLEX_TRITS_BRANCH);
+  return statement;
+}
+
+static CassStatement* ret_insert_address_edge_statement(const CassPrepared* prepared, const Transaction* transaction) {
```

Comment:
I dislike the way you implement `ret_insert_address_edge_statement`, `ret_insert_trunk_edge_statement`, `ret_insert_branch_edge_statement`, and `ret_insert_branch_edge_statement`. They have same logic and function invocations in common. Can you replace them with macros or inline functions?

Reply:
I merge `ret_insert_address_edge_statement`, `ret_insert_trunk_edge_statement` and `ret_insert_branch_edge_statement` to a single function `ret_insert_edge_statement` 
They are all for statement to Edge Table, only with different edge as the primary key.